_episodes/20-crop-raster-data.md

```diff
@@ -0,0 +1,218 @@
+from shapely.geometry import Polygon
+xlist= [630000, 629000, 638000, 639000, 634000, 630000]
+ylist = [5.804e6, 5.814e6, 5.816e6, 5.806e6, 5.803e6, 5.804e6]
+polygon_geom = Polygon(zip(xlist, ylist))
+polygon = gpd.GeoDataFrame(index=[0], crs=raster_clip.rio.crs, geometry=[polygon_geom])
+
+# Plot the polygon over raster
+fig, ax = plt.subplots()
+fig.set_size_inches((8,8))
+raster_clip.plot.imshow(ax=ax)
+polygon.plot(ax=ax, edgecolor='blue', alpha=0.6)
+
+# Crop and visualize
+raster_clip_polygon = raster_clip.rio.clip(polygon['geometry'], polygon.crs)
+raster_clip_polygon.plot.imshow(figsize=(8,8))
+~~~
+{: .language-python}
+
+<img src="../fig/20-crop-raster-crop-by-polygon-03.png" title="Crop raster by a polygon"  width="1024" style="display: block; margin: auto;" />
+
+ 
+> ## Exercise: Compare two ways of bounding box cropping
+> So far, we have learned two ways of cropping a raster: by a bounding box (using `clip_box`) and by a polygon (using `clip`). Technically, a bounding box is also a polygon. So what if we crop the original image directly with the polygon? For example:
+> ~~~
+> raster_clip_polygon2 = raster.rio.clip(polygon['geometry'], polygon.crs)
+> raster_clip_polygon2.plot.imshow()
+> ~~~
+> {: .language-python}
+> And try to compare the two methods:
+> - Do you have the same results?
+> - Do you have the same execution time?
+> - How would you choose the two methods in your daily work?
+>
+> > ## Solution
+> >
+> > The two methods give the same results, but cropping directly with a polygon is much slower.
+> >
+> > Therefore, if the AoI is much smaller than the original raster, it would be more efficient to first crop the raster with a bounding box, then crop with the actual AoI polygon.
+> {: .solution}
+{: .challenge}
+
+> ## Exercise: Select the raster data within crop fields
+> Can you select out all the crop fields from the raster data, using the `.shp` file in `data/crop_fields`? And also visualize your results.
+>
+> > ## Solution
+> >
+> > ~~~
+> > # Load the crop fields polygons 
+> > cf_boundary_crop = gpd.read_file("data/crop_fields/cf_boundary_crop.shp")
+> > # Crop
+> > raster_clip_fields = raster_clip.rio.clip(cf_boundary_crop['geometry'], cf_boundary_crop.crs)
+> > # Visualize
+> > raster_clip_fields.plot.imshow(figsize=(8,8))
+> > ~~~
+> > {: .language-python}
+> > <img src="../fig/20-crop-raster-crop-fields-solution-06.png" title="Raster cropped by crop fields" width="512" style="display: block; margin: auto;" />
+> {: .solution}
+{: .challenge}
+
+
+## Crop raster data with a geometry buffer
+
+It is not always the case that the AoI comes in the format of a polygon. Sometimes one would like to perform analysis around a (set of) point(s), or polyline(s). Using our AoI as an example, apart from the crop fields, one may want to also analyze raster data around the groundwater monitoring wells in the area. The location of the wells comes as point vector data in `data/groundwater_monitoring_well`.
```

Comment:
I think the sentence `Using our AoI as an example, apart from the crop fields, one may want to also analyze raster data around the groundwater monitoring wells in the area.` is complicated and difficult to understand. Can we rephrase it or break it into two sentences?

Reply:
```suggestion
It is not always the case that the AoI comes in the format of a polygon. Sometimes one would like to perform analysis around a (set of) point(s), or polyline(s). For example, in our AoI, there are also some groundwater monitoring wells coming as point vector data. One may also want to perform analysis around these wells. The location of the wells is stored in `data/groundwater_monitoring_well`.
```